doc: Adds TPF advanced_cluster migration guide & updates existing documentation and addresses remaining occurrences of PreviewProvider (#3584)
mongodb/terraform-provider-mongodbatlas: docs/guides/cluster-to-advanced-cluster-migration-guide.md

Due to its schema simplicity, `mongodbatlas_cluster` resource is unable to support most of the latest MongoDB Atlas features, such as [Multi-Cloud Clusters](https://www.mongodb.com/blog/post/introducing-multicloud-clusters-on-mongodb-atlas), [Asymmetric Sharding](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema), [Independent Scaling of Analytics Node Tiers](https://www.mongodb.com/blog/post/introducing-ability-independently-scale-atlas-analytics-node-tiers) and more.
On the other hand, not only does `mongodbatlas_advanced_cluster` cover everything that `mongodbatlas_cluster` can do, but it offers all existing MongoDB Atlas functionalities and will continue to do so going forward.

> Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` has npw been deprecated starting from provider version 2.0.0, and will soon be removed from the provider completely. Begin your planning now to ensure you're ready for this transition.

Reviewer: ```suggestion
Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` has now been deprecated starting from provider version 2.0.0, and will be removed from the provider in the following major versions. Begin your planning now to ensure you're ready for this transition.
```

Author: updated

> Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` has npw been deprecated starting from provider version 2.0.0, and will soon be removed from the provider completely. Begin your planning now to ensure you're ready for this transition.

Reviewer 2: ```suggestion
Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` is deprecated starting from provider version 2.0.0, and will soon be removed from the provider completely. Begin your planning now to ensure you're ready for this transition.
```

> ### What was the `mongodbatlas_advanced_cluster` Preview of MongoDB Atlas Provider 2.0.0?

Reviewer: ```suggestion
### What is the `mongodbatlas_advanced_cluster` Preview of MongoDB Atlas Provider 2.0.0 released in 1.X.0?
```

suggest to keep "is" but add the context that was released in 1.X. Can you please check which number is X?

Author: its 1.29 will update

> To make it easier to migrate to `mongodbatlas_advanced_cluster`, we decided to enable support for the [`moved` block](https://developer.hashicorp.com/terraform/language/moved). This functionality needed the resource to be implemented using the [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework), whereas our previous implementation [used the SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2). Given the considerable changes between the two frameworks and the breaking changes it causesd we decided to release a preview version of the `mongodbatlas_advanced_cluster`, in our provider versions 1.29.0 and later, usable under an environment variable and kept the existing implementation as-is to avoid breaking existing users. 

Reviewer: ```suggestion
To make it easier to migrate to `mongodbatlas_advanced_cluster`, we decided to enable support for the [`moved` block](https://developer.hashicorp.com/terraform/language/moved). This functionality needed the resource to be implemented using the [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework), whereas our previous implementation [used the SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2). Given the considerable changes between the two frameworks and the breaking changes it caused we decided to release a preview version of the `mongodbatlas_advanced_cluster`, in our provider versions 1.29.0 and later, usable under an environment variable and kept the existing implementation as-is to avoid breaking existing users. 
```

> To make it easier to migrate to `mongodbatlas_advanced_cluster`, we decided to enable support for the [`moved` block](https://developer.hashicorp.com/terraform/language/moved). This functionality needed the resource to be implemented using the [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework), whereas our previous implementation [used the SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2). Given the considerable changes between the two frameworks and the breaking changes it causesd we decided to release a preview version of the `mongodbatlas_advanced_cluster`, in our provider versions 1.29.0 and later, usable under an environment variable and kept the existing implementation as-is to avoid breaking existing users. 

Reviewer 2: ```suggestion
To make it easier to migrate to `mongodbatlas_advanced_cluster`, we enabled support for the [`moved` block](https://developer.hashicorp.com/terraform/language/moved). This functionality requires you to implement the resource using the [Terraform Plugin Framework](https://developer.hashicorp.com/terraform/plugin/framework). However, our previous implementation [used the SDKv2](https://developer.hashicorp.com/terraform/plugin/sdkv2). Given the considerable changes between the two frameworks and the breaking changes it causes, we releases a preview version of the `mongodbatlas_advanced_cluster`, in our provider versions 1.29.0 and later, which you can use under an environment variable and keep the existing implementation as-is to avoid breaking existing implementations. 
```

Now with the release of MongoDB Atlas Provider 2.0.0, only the new version remains and the environment variable is no longer needed.

> More information about the new version of `mongodbatlas_advanced_cluster` can be found in the [resource documentation page](../resources/advanced_cluster).

Reviewer: ```suggestion
More information about `mongodbatlas_advanced_cluster` can be found in the [resource documentation page](../resources/advanced_cluster).
```

> More information about the new version of `mongodbatlas_advanced_cluster` can be found in the [resource documentation page](../resources/advanced_cluster).

Reviewer 2: ```suggestion
To learn more about the new version of `mongodbatlas_advanced_cluster`, see the [resource documentation page](../resources/advanced_cluster).
```

> More information about the new version of `mongodbatlas_advanced_cluster` can be found in the [resource documentation page](../resources/advanced_cluster).

Author: ```suggestion
To learn more about `mongodbatlas_advanced_cluster`, see the [resource documentation page](../resources/advanced_cluster).
```

Due to its schema simplicity, `mongodbatlas_cluster` resource is unable to support most of the latest MongoDB Atlas features, such as [Multi-Cloud Clusters](https://www.mongodb.com/blog/post/introducing-multicloud-clusters-on-mongodb-atlas), [Asymmetric Sharding](https://registry.terraform.io/providers/mongodb/mongodbatlas/latest/docs/guides/advanced-cluster-new-sharding-schema), [Independent Scaling of Analytics Node Tiers](https://www.mongodb.com/blog/post/introducing-ability-independently-scale-atlas-analytics-node-tiers) and more.
On the other hand, not only does `mongodbatlas_advanced_cluster` cover everything that `mongodbatlas_cluster` can do, but it offers all existing MongoDB Atlas functionalities and will continue to do so going forward.

> Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` is deprecated starting from provider version 2.0.0, and will soon be removed from the provider in the following major versions. Begin your planning now to ensure you're ready for this transition.

Reviewer: ```suggestion
Having that in mind, to access all the latest functionalities and stay up to date with our best offering we recommend you to start planning your move to `mongodbatlas_advanced_cluster`. To maintain our focus on enhancing the overall experience with `mongodbatlas_advanced_cluster`, `mongodbatlas_cluster` is deprecated starting from provider version 2.0.0, and will be removed from the provider in the following major versions. Begin your planning now to ensure you're ready for this transition.
```

please let's remove any temporal reference here. Also "soon" might be scaring

**Prerequisites:**
 - Terraform version 1.8 or later is required, more information in the [State Move page](https://developer.hashicorp.com/terraform/plugin/framework/resources/state-move).
 - MongoDB Atlas Provider version 2.0 or later is required. 

>      - Note: If using version 1.29 or later, you can leverage the preview of the new schema. To learn more, please review our upgrade guides for the version you are using.

Reviewer 3: this is contradictory, if we state that 2.0 or above is required, saying that "if using version 1.29 or later..." does not make sense

Author: hmm maybe we can phrase it differently? I thought we should continue to mention the preview for now for users who are not yet ready to completely migrate to 2.0 because of all the changes but still may want to "try out" the new cluster schema. Given that we saw low adoption of preview anyway, this may help to motivate users to try the new schema now. WDYT? lmk if you have any suggestions

Reviewer 3: Makes more sense now that you explained it, it's a tricky one. 
I would say that it would be clearer if the note said something like "If you are migrating to mongodbatlas_advanced_cluster in a version between 1.29 and 2.0.0, you can leverage..."
What do you think?

Author: not sure, we can wait for the team to chime in

>      - Note: If using version 1.29 or later, you can leverage the preview of the new schema. To learn more, please review our upgrade guides for the version you are using.

Reviewer 2: ```suggestion
     - Note: If you use version 1.29 or later, you can leverage the preview of the new schema. To learn more, please review our upgrade guides for the version you are using.
```

The basic experience when using the `moved` block is as follows:
1. Before starting, run `terraform plan` to make sure that there are no planned changes.
2. Add the `mongodbatlas_advanced_cluster` resource definition.

>   - You can use the [Atlas CLI plugin](https://github.com/mongodb-labs/atlas-cli-plugin-terraform) to generate the `mongodbatlas_advanced_cluster` resource definition. This is the recommended method as it will generate a clean configuration while keeping the original Terraform expressions. Please be aware of the [plugin limitations](https://github.com/mongodb-labs/atlas-cli-plugin-terraform#limitations), see the [section below](#alternatives-to-using-the-mongodb-atlas-cli-plugin-to-generate-the-mongodbatlas_advanced_cluster-resource-definition) for the available alternatives.

Author: @lantoli can you confirm if you'll be updating the TPF GA changes & docs in the CLI plugin repo as well?

Reviewer 4: @maastha yes, I'll update the plugin repo docs

3. Comment out or delete the `mongodbatlas_cluster` resource definition.
4. Update the references from your previous cluster resource: `mongodbatlas_cluster.this.XXXX` to the new `mongodbatlas_advanced_cluster.this.XXX`.
   - Double check [output-changes](#output-changes) to ensure the underlying configuration stays unchanged.

>    - If you are using output variables that use the new resource `mongodbatlas_advanced_cluster.this`, the plan output can be more verbose than expected (extra `Note: Objects have changed outside of Terraform` section). Consider adding/updating output variables only **after** performing the move (see more in the [Github Issue](https://github.com/hashicorp/terraform/issues/36796)).

Reviewer 2: ```suggestion
   - If you use output variables that use the new resource `mongodbatlas_advanced_cluster.this`, the plan output can be more verbose than expected (extra `Note: Objects have changed outside of Terraform` section). Consider adding or updating output variables only **after** performing the move (see more in the [Github Issue](https://github.com/hashicorp/terraform/issues/36796)).
```
